R/gadgets.R

```diff
@@ -0,0 +1,58 @@
+#' Open dataset selector
+#'
+#' This function launches a shiny gadget which allows you to navigate your Crunch
+#' projects and datasets. This is useful if you can't remember a dataset's project
+#' and also saves typing long dataset names.
+#'
+#' @inheritParams listDatasets
+#' @return A `listDatasets()` call is pasted into your RStudio session`
+#' @export
+listDatasetGadget <- function(kind=c("active", "all", "archived"),
+    refresh = FALSE){
+    projects <- c("Personal Project", names(projects()))
+    personal_datasets <- listDatasets(kind = kind, refresh = refresh)
+
+    ui <- miniUI::miniPage(
+        miniUI::gadgetTitleBar("Select Dataset"),
+        miniUI::miniContentPanel(
+            shiny::column(width = 3,
+                shiny::selectInput("project",
+                    "Select Project",
+                    projects)),
+            shiny::column(width = 3,
+                shiny::uiOutput("dataset")
+            ),
+            shiny::column(width = 3,
+                shiny::textInput("ds_name", "Object Name (Optional)")
+            )
+        )
+    )
+
+    server <- function(input, output, session) {
+        output$dataset <- shiny::renderUI({
+            if (input$project == "Personal Project") {
+                selections <- personal_datasets
+            } else {
+                selections <- listDatasets(project = input$project, kind = kind, refresh = refresh)
+            }
+            shiny::selectInput("dataset",
+                "Select Dataset",
+                selections)
+        })
+        shiny::observeEvent(input$done, {
+            assignment <- ifelse(nchar(input$ds_name) > 0,
+                paste0(input$ds_name, " <- "),
+                "")
+            if (input$project == "Personal Project") {
+                code <- paste0(assignment, "loadDataset('", input$dataset,"')")
+            } else {
+                code <- paste0(assignment, "loadDataset('",
+                    input$dataset,
+                    "', project = '",
+                    input$project, "')" )
```

Comment:
This will result in invalid syntax if either dataset or project name has an apostrophe in it. 

Consider factoring out this syntax construction to a function that you can unit test.

Reply:
Replying here to the general review comment about testing shiny apps. 

It seems like the best method for testing shiny apps in general is the shinytest package (https://github.com/rstudio/shinytest) which allows you to record actions taken in a shiny session. Unfortunately this currently doesn't support shiny apps which are launched from a project function which makes this not that useful for this case. 

What I've done is pulled as much logic as I can out of the shiny function and tested it using our existing approach. I considered doing something like mocking the input list and comparing the output of the server function to a baseline, but the shiny environment which is returned by the render functions is a bit complicated to make that useful. Moreover it might cause spurious test failures if the shiny library changes. 

So there are two basic ways to test the shiny logic: 
1) Extend `shinytest` to accommodate apps launched from functions
2) Pull the server and ui elements into a folder based app and test that separately. 

In this case it might not be worth setting that up as the shiny logic is pretty straightforward, but I'm happy to look into it further if we think it's a priority.